### ISSUE TITLE
setting one configuration parameter in model's file erases all other configuration parameters set at directory and project levels

### ISSUE THREAD

**Issue by fetanchaud** [fishtown-analytics/dbt#1343]
## Issue

Setting one configuration parameter in model's file erases all other configuration parameters set at directory and project levels

### Issue description

Example files : 

#### Project : 

```
name: 'materialization_config'
version: '1.0'

profile: 'dwh'

quoting:
    identifier: false
    schema: false

source-paths: ["models"]
analysis-paths: ["analysis"] 
test-paths: ["tests"]
data-paths: ["data"]
macro-paths: ["macros"]

target-path: "target"
clean-targets:       
    - "target"
    - "dbt_modules"

models:
    materialization_config:

        materialized: my_mat
        my_project_param: a_project_param
        
        my_directory:
            my_dir_param: a_dir_param
        
            my_model:
                my_project_file_model_param: a_project_file_param
```

#### Materialization : 

```

{% materialization my_mat, default -%}
    {%- set my_project_param = config.get('my_project_param') -%}
    {%- set my_dir_param = config.get('my_dir_param') -%}
    {%- set my_project_file_model_param = config.get('my_project_file_model_param') -%}
    {%- set my_model_file_param = config.get('my_model_file_param') -%}    
    
    {{ log('project param:' ~ my_project_param, info=True)}}
    {{ log('directory param:' ~ my_dir_param, info=True)}}
    {{ log('project file model param:' ~ my_project_file_model_param, info=True)}}
    {{ log('model file param:' ~ my_model_file_param, info=True)}}    
    
    
     {%- call statement('main') -%}    
       select 'Hello world' as message
     {%- endcall -%}
    
{%- endmaterialization %}

```

#### model's file : 
```

{{
    config(
        my_model_file_param='a_model_file_param'
    )
}}

```

### Results

```

$ dbt run
Found 1 models, 0 tests, 0 archives, 0 analyses, 96 macros, 2 operations, 0 seed files

10:58:25 | Concurrency: 4 threads (target='dev')
10:58:25 |
10:58:25 | 1 of 1 START my_mat model public.my_model............................ [RUN]
project param:None
directory param:None
project file model param:None
model file param:a_model_file_param
10:58:26 | 1 of 1 OK created my_mat model public.my_model....................... [SELECT 1 in 0.11s]
10:58:26 |
10:58:26 | Finished running 1 my_mat models in 1.14s.


```

I thought configuration parameters were inherited and maybe overriden on the way if needed,
but it seems that eventually overriding one parameter erases all others' values.

Is it expected behaviour ? 

Thank you !

### System information

windows 7
dbt 0.12.2
python 3.7.1

**Comment by drewbanin** [comment]
Hey @fetanchaud - I agree with your assessment that the configs specified in `dbt_project.yml` aren't coming through in the custom materialization. I don't think this has to do with the single configuration parameter defined in the `{{ config(...) }}` block however. Even if you remove that config line, i think you'll see the same behavior. (Note: dbt will not execute empty files, so just put a comment in there to test this out).

Instead, I think what's happening is that dbt only includes "known" configs specified in the `dbt_project.yml` file. This list of configs is defined [here](https://github.com/fishtown-analytics/dbt/blob/dev/stephen-girard/core/dbt/parser/source_config.py#L8-L19). When dbt encounters these configs, it will apply them to any applicable models on a hierarchical basis as it sounds like you're expecting.

The big problem is that when dbt encounters a config like `my_project_param`, it doesn't know if that's a directory name or a known config! This is sort of unavoidable given the current ambiguity in dbt's model configuration syntax.

I'd be open to exploring some mechanism for specifying configs dynamically in dbt_project.yml `models:` config, though in the meantime, I'd recommend specifying these configurations in a `{{ config(...) }}` block. There are a couple of mechanisms at your disposal that might make this easier for you:

**1. Use macros to wrap config()**

If you find yourself repeating the same config calls, try encapsulating them inside of a macro!

```sql
-- macros/standard_config.sql
{% macro standard_config(my_model_file_param) %}
  {{
    config(
      my_project_param='a_project_param',
      my_dir_param='a_dir_param',
      my_model_file_param=my_model_file_param
  ) }}
{% endmacro %}
```

```sql
-- models/example/my_model.sql
{{ standard_config(my_model_file_param='a_model_file_param') }}
....
```

**2. Use vars to supply configuration values hierarchically**

Rather than specifying configs directly in the `dbt_project.yml` file, you can instead use dbt's [variables](https://docs.getdbt.com/docs/using-variables) for a similar purpose. Using variables, your example above would look something like:


```yml
# dbt_project.yml

...

models:
    materialization_config:

        materialized: my_mat
        vars:
            my_project_param: a_project_param
        
        my_directory:
            vars:
               my_dir_param: a_dir_param
        
            my_model:
                vars:
                    my_project_file_model_param: a_project_file_param
```

```sql
-- macros/mat.sql

{% materialization my_mat, default -%}
    {%- set my_project_param = var('my_project_param') -%}
    {%- set my_dir_param = var('my_dir_param') -%}
    {%- set my_project_file_model_param = var('my_project_file_model_param') -%}
 
    {#-- this would still unfortunately need to be a config. Vars can't be specified inside a model #}
    {%- set my_model_file_param = config.get('my_model_file_param') -%}    
    
    {{ log('project param:' ~ my_project_param, info=True)}}
    {{ log('directory param:' ~ my_dir_param, info=True)}}
    {{ log('project file model param:' ~ my_project_file_model_param, info=True)}}
    {{ log('model file param:' ~ my_model_file_param, info=True)}}    
    
    
     {%- call statement('main') -%}    
       select 'Hello world' as message
     {%- endcall -%}
    
{%- endmaterialization %}
```

Finally, note that you can use both of these approaches in tandem! Hopefully between both of these options, you'll be able to find a workable solution.

Hope this helps!

**Comment by fetanchaud** [comment]
Thank you @drewbanin,

I used the vars in my first persistent staging area approach (based on a single macro call in the model's file),
but when I switched to materialization, I converted these vars in config params, and, as vars and already known config params were inherited, I took it to granted. Now I get the picture. Thank you  !
If you don't mind I close this non-issue.

Best,
fabrice etanchaud
